text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,127 @@
+
+# Design
+
+## Outputs API
+Only outputs with the `OBS_OUTPUT_SERVICE` flag are concerned.
+
+Adding to `obs_output_info`:
+- `const char *protocols`: protocols separated with a semi-colon (ex: `"RTMP;RTMPS"`), required to register the output. The string used to identify the protocol must be its officialy/widely used acronym, those acronyms are usually uppercased.
+
+Adding to this API these functions:
+- `const char *obs_output_get_protocols(const obs_output_t *output)`: returns protocols supported by the output.
+- `bool obs_is_output_protocol_registered(const char *protocol)`: return true if an output with the protocol is registered.
+- `bool obs_enum_output_service_types(const char *protocol, size_t idx, const char **id)`: enumerate all outputs types compatible with the given protocol.
+- `bool obs_enum_output_protocols(size_t idx, const char **protocol)`: enumerate all registered protocol.
+- `const char *obs_get_output_supported_video_codecs(const char *id)`: return compatible video codecs of the given output id.
+- `const char *obs_get_output_supported_audio_codecs(const char *id)`: return compatible audio codecs of the given output id.
+
+## Services API
+Since a streaming service may not accept all the codecs usable by a protocol, adding a way to set supported codecs is required.
+
+Adding to `obs_service_info`:
+- `const char *(*get_protocol)(void *data)`: return the protocol used by the service. RFC 39 will allow multi-protocol services in the future.
+- `const char **(*get_supported_video_codecs)(void *data)`: video codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+- `const char **(*get_supported_audio_codecs)(void *data)`: audio codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+
+Adding to this API these functions:
+- `const char *obs_service_get_protocol(const obs_service_t *service)`: return the protocol used by the service.
+- `const char **obs_service_get_supported_video_codecs(const obs_service_t *service)`: return video codecs compatible with the service.
+- `const char **obs_service_get_supported_audio_codecs(const obs_service_t *service)`: return audio codecs compatible with the service.
+
+### Services and connection informations
+
+Depending on the protocol, the service object should provide various types of connection details (e.g., server URL, stream key, username, password).
+
+Currently, `get_key` can provide a stream id (SRT) or an encryption passphrase (RIST) rather than a stream key.
+
+Rather than adding a getter for each possible type of information, a getter where we can choose which type of information we want should be implemented.
+
+Types of connection details will be defined by an enum with only even number values. Odd number values will be reserved for potential third-party protocols (RFC 39).
+
+List of types:
+```c
+enum obs_service_connect_info {
```

Comment:
Sorry to nitpick, but given that this info is specifically associated with a protocol, perhaps `obs_service_protocol_info` is a more suitable name.

Reply:
I choose `connect` because this is an information used to connect to the ingest.

`protocol` would indicate that it's information about the protocol itself and not the connection to ingest.

Reply:
A reply to allow resolution of this comment